server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -8,32 +8,193 @@
+    private volatile Scheduler.Cancellable scheduledFuture;
+    private final ThreadPool threadPool;
+    private final ClusterService clusterService;
+    private final WorkloadManagementSettings workloadManagementSettings;
+    private Set<QueryGroup> activeQueryGroups;
+    private final Set<QueryGroup> deletedQueryGroups;
+    private final NodeDuressTrackers nodeDuressTrackers;
+
+    public QueryGroupService(
+        QueryGroupTaskCancellationService taskCancellationService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        WorkloadManagementSettings workloadManagementSettings
+    ) {
+
+        this(
+            taskCancellationService,
+            clusterService,
+            threadPool,
+            workloadManagementSettings,
+            new NodeDuressTrackers(
+                Map.of(
+                    ResourceType.CPU,
+                    new NodeDuressTracker(
+                        () -> workloadManagementSettings.getNodeLevelCpuCancellationThreshold() < ProcessProbe.getInstance()
+                            .getProcessCpuPercent() / 100.0,
+                        workloadManagementSettings::getDuressStreak
+                    ),
+                    ResourceType.MEMORY,
+                    new NodeDuressTracker(
+                        () -> workloadManagementSettings.getNodeLevelMemoryCancellationThreshold() <= JvmStats.jvmStats()
+                            .getMem()
+                            .getHeapUsedPercent() / 100.0,
+                        workloadManagementSettings::getDuressStreak
+                    )
+                )
+            ),
+            new HashMap<>(),
+            new HashSet<>(),
+            new HashSet<>()
+        );
+    }
+
+    public QueryGroupService(
+        QueryGroupTaskCancellationService taskCancellationService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        WorkloadManagementSettings workloadManagementSettings,
+        NodeDuressTrackers nodeDuressTrackers,
+        Map<String, QueryGroupState> stateMap,
+        Set<QueryGroup> activeQueryGroups,
+        Set<QueryGroup> deletedQueryGroups
+    ) {
+        this.taskCancellationService = taskCancellationService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.nodeDuressTrackers = nodeDuressTrackers;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        activeQueryGroups.forEach(queryGroup -> stateMap.putIfAbsent(queryGroup.get_id(), new QueryGroupState()));
+        this.queryGroupStateMap = stateMap;
+        this.queryGroupStateMap.put(QueryGroupTask.DEFAULT_QUERY_GROUP_ID_SUPPLIER.get(), new QueryGroupState());
+        taskCancellationService.setQueryGroupStateMapAccessor(this::getQueryGroupState);
```

Comment:
I am wondering if there is better way than using this setter for supplying the `QueryGroupStateMapAccessor` to `QueryGroupTaskCancellationService`? Injecting `QueryGroupService` in `QueryGroupTaskCancellationService` will probably result in circular dependency during initialization.

Reply:
IMO injecting QueryGroupService is unnecessary since `QueryGroupTaskCancellationService ` doesn't need the whole object. Maybe something like the following could be fine 

**QueryGroupsStateMapAccessor**: responsible for supplying the composite stateMap or queryGroup level stateMap

now  both `QueryGroupService` and `QueryGroupTaskCancellationService` consume this.

What do you think ?

Reply:
Yeah, that should work better!